### ISSUE TITLE
RL agent validation: action mapping alignment & always-HOLD policy diagnostics

### ISSUE THREAD

**Issue by SefaGH** [SefaGH/bearish-alpha-bot#413]
## Summary

The RL training/validation pipeline is now running end-to-end with GEMMA + 82 features and the new RL workflows. However, the current trained RL checkpoint appears to collapse to an **“always HOLD / no-trade” policy** during validation.

At the same time, there is a potential mismatch between how action indices are **interpreted by the environment** vs. how they are **logged/treated in the agent**, and the validation report JSON format is not fully aligned with the summary script.

This issue is to:
1. Normalize and verify the action mapping end-to-end.
2. Improve the validation report / summary for the RL agent.
3. Add minimal diagnostics to avoid silent “always HOLD / no-trade” policies.

---

### 1. Validation report: 100% HOLD, reward and PnL = 0

From the latest `rl_validation_report.json`:

```json
{
  "dataset": "data/training/BTC_USDT_USDT_1h_val.npz",
  "checkpoint": "data/checkpoints/rl_agent_final.pth",
  "symbol": "BTC/USDT:USDT",
  "timeframe": "1h",
  "generated_at": "2025-11-18T22:37:03.667058Z",
  "metrics": {
    "steps": 899,
    "total_reward": 0.0,
    "final_pnl": 0.0,
    "action_counts": {
      "0": 899,
      "1": 0,
      "2": 0
    },
    "action_distribution": {
      "0": 1.0,
      "1": 0.0,
      "2": 0.0
    },
    "q_values": {
      "mean": [
        -0.01587575301527977,
        -0.016085676848888397,
        -0.01613043062388897
      ],
      "std": [
        1.2479722499847412e-07,
        1.0803341865539551e-07,
        8.940696716308594e-08
      ],
      "min": [
        -0.01587562821805477,
        -0.016085784882307053,
        -0.016130341216921806
      ],
      "max": [
        -0.01587562821805477,
        -0.016085784882307053,
        -0.016130341216921806
      ]
    },
    "probabilities": {
      "mean": [
        0.33338457345962524,
        0.33331361413002014,
        0.33330056071281433
      ],
      "std": [
        3.8743019104003906e-07,
        1.3113021850585938e-06,
        5.066394805908203e-07
      ],
      "min": [
        0.3333849608898163,
        0.3333149254322052,
        0.33330005407333374
      ],
      "max": [
        0.3333849608898163,
        0.3333149254322052,
        0.33330005407333374
      ]
    }
  }
}
```

Key points:

- `action_counts`: **0 → 899, 1 → 0, 2 → 0**
- `total_reward = 0.0`, `final_pnl = 0.0`
- Q-values:
  - means around `[-0.0159, -0.0161, -0.0161]`
  - std ~ `1e-7` (almost constant across all steps)
- Probabilities:
  - ~ `[0.33338, 0.33331, 0.33330]`
  - std ~ `1e-6` (almost constant)

So the trained policy at validation time:

- emits **almost identical Q and probability vectors for all states**, and
- yields a deterministic argmax that always picks action index `0`.

Given how the environment interprets actions, this leads to **no trades at all**, and thus 0 reward and 0 PnL.

---

### 2. Environment action mapping

`src/ml/rl_trading_env.py`:

```python
class RLTradingEnv:
    def __init__(..., initial_balance: float = 10000.0, fee: float = 0.0006):
        ...
        self.state_dim = len(features_df.columns)
        self.action_dim = 3  # 0: Hold, 1: Buy, 2: Sell
        ...

    def step(self, action: int):
        ...
        # --- Execute Action ---
        if action == 1:  # Buy
            ...
        elif action == 2:  # Sell
            ...
        ...
```

So the **canonical mapping in the environment** is:

- `0`: Hold
- `1`: Buy
- `2`: Sell

---

### 3. Agent action semantics and logging

`src/ml/reinforcement_learning.py`:

```python
def get_action_with_meta(...):
    ...
    if effective_training and random.random() < self.epsilon:
        action = random.randrange(self.action_size)
        meta['exploration'] = True
        meta['probabilities'] = None
        logger.debug(
            f"🤖 [RL-ACT] Exploration: Selected random action -> "
            f"{['BUY', 'HOLD', 'SELL'][action]}"
        )
        return action, meta
    ...
    probabilities = torch.softmax(adjusted_q_values, dim=1).squeeze().cpu().numpy()
    best_action = int(np.argmax(probabilities))
    best_prob = float(probabilities[best_action])
    ...
    return best_action, meta
```

Notes:

- The logging string uses `['BUY', 'HOLD', 'SELL'][action]`, which implies **0=BUY,1=HOLD,2=SELL** for logs.
- But the environment uses `0=HOLD,1=BUY,2=SELL`.
- Validation calls the agent like this:

  ```python
  action, meta = agent.get_action_with_meta(state, training=False)
  next_state, reward, done, info = env.step(action)
  ```

So there is at least a **semantic/logging inconsistency**, and possibly a deeper mismatch if any part of the code was written assuming `0=BUY`.

In validation we see:

- `action_counts["0"]=899`
- but in the environment, action `0` means **HOLD**,
- hence: no trades, no reward, no PnL.

---

### 4. Validation report vs. summary script

`scripts/validate_rl_model.py` builds the report as:

```python
metrics = evaluate_agent(agent, env, args.max_steps)
payload = {
    "dataset": str(dataset_path),
    "checkpoint": str(checkpoint) if checkpoint else None,
    "symbol": meta.get("symbol"),
    "timeframe": meta.get("timeframe"),
    "generated_at": datetime.utcnow().isoformat() + "Z",
    "metrics": metrics,
}
```

The actual fields live under `payload["metrics"]`:

- `metrics["total_reward"]`
- `metrics["final_pnl"]`
- `metrics["action_counts"]`
- `metrics["action_distribution"]`
- `metrics["q_values"]` / `metrics["probabilities"]`

The summary step (`scripts/summarize_rl_validation.py` in the workflow) was initially reading different top-level keys (`total_reward`, `pnl_total`, etc.), so it showed `N/A` and `{}` even though the underlying JSON holds the correct metrics.

This has been partially addressed by adjusting the summary script to read from `data["metrics"]`, but the contract between `validate_rl_model.py` and the summary script should be made explicit and stable.

---

## Tasks for Copilot / Implementation Plan

Please create a PR that addresses the following tasks. Use small, localized changes with clear logging where possible, and keep behavior changes minimal but explicit.

### Task 1 — Normalize and verify action mapping end-to-end

1. **Confirm and document the canonical mapping** across the RL stack:

   - **Canonical mapping (target):**

     - `0` → `HOLD`
     - `1` → `BUY`
     - `2` → `SELL`

   - Update any comments/docstrings that disagree with this, particularly in:
     - `src/ml/rl_trading_env.py`
     - `src/ml/reinforcement_learning.py`
     - `scripts/train_rl_agent.py` (if it labels actions somewhere)
     - Any other RL-related scripts that assume a particular order.

2. **Fix logging names to match mapping:**

   - In `TradingRLAgent.get_action_with_meta`, change the debug string:

     ```python
     ['BUY', 'HOLD', 'SELL'][action]
     ```

     to a list that matches the canonical mapping:

     ```python
     ['HOLD', 'BUY', 'SELL'][action]
     ```

   - Audit any other hard-coded action-name arrays and align them.

3. **Add a small self-check / unit test:**

   - Option A: A simple unit test that asserts the intended mapping is documented and used consistently (e.g. a test that uses a stub env and ensures action `0` does not trigger a trade).
   - Option B: At least log clearly, once at agent/env init, what each index means, with a single source of truth constant (e.g., `ACTION_LABELS = ['HOLD','BUY','SELL']`).

4. **Ensure training and validation share the same mapping:**

   - `scripts/train_rl_agent.py` must use the same `RLTradingEnv` and not post-process action indices in a conflicting way.
   - If there is any mapping layer between agent and env, make it explicit and consistent.

---

### Task 2 — Stabilize validation report format and summary

1. **Make `scripts/validate_rl_model.py`’s output schema explicit:**

   - Document in the module docstring and/or inline comments:

     ```python
     payload = {
       "dataset": ...,
       "checkpoint": ...,
       "symbol": ...,
       "timeframe": ...,
       "generated_at": ...,
       "metrics": {
         "steps": int,
         "total_reward": float,
         "final_pnl": float,
         "action_counts": {str: int},
         "action_distribution": {str: float},
         "q_values": {
           "mean": [float, float, float],
           "std": [...],
           "min": [...],
           "max": [...]
         },
         "probabilities": { ... }
       }
     }
     ```

   - Ensure these keys are always present, even if some metrics are `None` or empty (e.g., when no Q history exists).

2. **Make `scripts/summarize_rl_validation.py` consume this schema:**

   - Read from `data["metrics"]` instead of top-level keys.
   - Print at least:

     - `Total Reward: metrics["total_reward"]`
     - `Final PnL:    metrics["final_pnl"]`
     - `Action Distribution: metrics["action_distribution"]`
     - `Q-Value Stats: metrics["q_values"]` (or a summarized view)

3. **Optionally add a simple “health check” in the summary:**

   - If `action_distribution["0"] > 0.99` and `total_reward` and `final_pnl` are ~0:
     - Print a clear warning such as:

       > “WARNING: RL policy appears to be degenerate (>=99% HOLD, ~0 reward/PnL). Consider inspecting training logs, Q std, and reward scale.”

   - This is just logging, not a hard failure, but it makes this failure mode obvious in CI.

---

### Task 3 — (Minimal) diagnostics for degenerate Q / always-HOLD policies

Without doing a full RL hyperparameter overhaul yet, add minimal diagnostics that help detect the current failure mode:

1. **Q-Std diagnostics:**

   - We’ve already seen from `inspect_replay_and_td.py` and validation:

     - Train: average Q std ≈ `1e-4`
     - Validation: per-action std ≈ `1e-7`

   - Add a log in either:
     - training summary (`rl_training_summary.json`), or
     - validation report (already has `q_values.std`),
     that clearly captures whether Q-std is below some tiny threshold (e.g. `1e-5`) and log a warning.

2. **Hold-rate diagnostics:**

   - In validation, we already have `action_distribution`.
   - Add a warning log if HOLD rate is extremely high (e.g. `> 0.95`) or BUY/SELL are essentially never used.

No behavior change in the policy itself is required as part of this issue; focus is on **correct mapping + better observability**. Hyperparameter tuning (reward_scale, learning_rate, hold_confidence_threshold, etc.) can be a follow-up issue.

---

## Acceptance Criteria

- [ ] Action index mapping is clearly documented and consistent across:
  - `RLTradingEnv`
  - `TradingRLAgent` (including log messages)
  - Training and validation scripts.
- [ ] Validation report JSON (`rl_validation_report.json`) contains the documented schema, with `metrics.total_reward`, `metrics.final_pnl`, `metrics.action_distribution`, and `metrics.q_values` present.
- [ ] `scripts/summarize_rl_validation.py` reads from `metrics` and prints the correct values in CI logs.
- [ ] When the RL policy collapses to “always HOLD / no-trade” (like the current checkpoint), the validation summary emits at least one clear warning (log or stdout) so this is obvious without manually inspecting the JSON.
- [ ] Optional: a small unit test confirming that the canonical mapping `0=HOLD, 1=BUY, 2=SELL` is respected by the RL env and agent wiring.

Please keep changes well-scoped and add/adjust tests where practical.
